File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.py
class Solution(object):
    def canplace(self,position,mid,m):
        prev = position[0]
        cnt = 1
        for i in range(1,len(position)):
            if position[i]-prev>=mid:
                cnt+=1
                prev = position[i]
        return cnt>=m

    def maxDistance(self, position, m):
        position.sort()
        l = 1
        ans = 0
        r=(position[-1]-position[0])/(m-1)
        while l<=r:
            mid = l+(r-l)/2
            if self.canplace(position,mid,m):
                ans = mid
                l = mid+1
            else:
                r=mid-1
        return ans

File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.cpp
class Solution {
    bool canplace(vector<int>&position,int m,int mid){
        int prev = position[0];
        int cnt = 1;
        for(int i=1;i<position.size();i++){
            if(position[i]-prev>=mid){
                prev = position[i];
                cnt++;
            }
        }
        return cnt>=m;
    }
public:
    int maxDistance(vector<int>& position, int m) {
        sort(position.begin(),position.end());
        int n = position.size();
       
           int l = 1;
           int r =(position.back() - position[0]) / (m - 1);
            int ans = 1;
        while(l<=r){
            int mid = l+(r-l)/2;
            if(canplace(position,m,mid)){
                ans = mid;
                l=mid+1;
            }
            else r=mid-1;
        }
        return ans;
    }
};